apm-agent-plugins/apm-vertx/apm-vertx4-plugin/src/main/java/co/elastic/apm/agent/vertx/v4/webclient/WebClientInstrumentation.java

```diff
@@ -0,0 +1,190 @@
+
+        @Override
+        public ElementMatcher<? super TypeDescription> getTypeMatcher() {
+            return named("io.vertx.ext.web.client.impl.HttpContext");
+        }
+
+    }
+
+    /**
+     * Instruments TODO
+     */
+    public static class HttpContextPrepareRequestInstrumentation extends HttpContextInstrumentation {
+
+        @Override
+        public ElementMatcher<? super MethodDescription> getMethodMatcher() {
+            return named("prepareRequest");
+        }
+
+        @Override
+        public String getAdviceClassName() {
+            return "co.elastic.apm.agent.vertx.v4.webclient.WebClientInstrumentation$HttpContextPrepareRequestInstrumentation$HttpContextPrepareRequestAdvice";
+        }
+
+        public static class HttpContextPrepareRequestAdvice {
+
+
+            @Advice.OnMethodExit(suppress = Throwable.class, inline = false)
+            public static void prepareRequest(@Advice.This HttpContext<?> httpContext) {
+                AbstractSpan<?> activeSpan = tracer.getActive();
+                if (null != activeSpan) {
+                    activeSpan.incrementReferences();
+                    httpContext.set(WEB_CLIENT_PARENT_SPAN_KEY, activeSpan);
+                }
+            }
+        }
+    }
+
+    /**
+     * Instruments TODO
+     */
+    public static class HttpContextSendRequestInstrumentation extends HttpContextInstrumentation {
+
+        @Override
+        public ElementMatcher<? super MethodDescription> getMethodMatcher() {
+            return named("sendRequest").and(takesArgument(0, named("io.vertx.core.http.HttpClientRequest")));
+        }
+
+        @Override
+        public String getAdviceClassName() {
+            return "co.elastic.apm.agent.vertx.v4.webclient.WebClientInstrumentation$HttpContextSendRequestInstrumentation$HttpContextSendRequestAdvice";
+        }
+
+        public static class HttpContextSendRequestAdvice {
+
+
+            @Advice.OnMethodEnter(suppress = Throwable.class, inline = false)
+            public static void sendRequest(@Advice.This HttpContext<?> httpContext, @Advice.Argument(value = 0) HttpClientRequest request, @Advice.FieldValue(value = "context") Context vertxContext) {
+                Object spanObj = httpContext.get(WEB_CLIENT_PARENT_SPAN_KEY);
+                boolean decrementParentReferences = true;
+                if (null == spanObj) {
+                    spanObj = vertxContext.getLocal(AbstractVertxWebHelper.CONTEXT_TRANSACTION_KEY);
+                    decrementParentReferences = false;
+                }
+
+                if (spanObj instanceof AbstractSpan) {
+                    AbstractSpan<?> parent = (AbstractSpan<?>) spanObj;
+                    webClientHelper.startSpan(parent, httpContext, request);
+                    if (decrementParentReferences) {
+                        parent.decrementReferences();
```

Comment:
Conceptually, reference increment should be coupled with the insertion into `HttpContext` and reference decrement should be coupled with removal (which is actually setting the key to `null` as there doesn't seem to be a `remove` API).
Removing from the map is important - if we decrement references and the span object gets recycled, the `HttpContex` will point to a span with arbitrary contents.
We need to pay attention to both `WEB_CLIENT_PARENT_SPAN_KEY` and `WEB_CLIENT_SPAN_KEY`.